### PR TITLE
Amls 4335 implement flow

### PR DESCRIPTION
**Pull request by ajchilts**
Implement flow for approval check and ensure both fit and proper and approval values get set correctly/ 

## Related / Dependant PRs?

N/A

## How Has This Been Tested?

Unit tests and manual testing given the scenarios. Regression automation also run. 

## Checklist

- [ ] Breaking change (fix or feature that would cause existing functionality to change).
- [x] Build passing.
- [x] Unit tests have full coverage.
- [ ] Unit test approach and implementation has been reviewed with QA (testers).
- [x] Requires acceptance test run.
- [x] Appropriate labels added.

### CODE REVIEW

File: app/controllers/responsiblepeople/FitAndProperController.scala [hunk 1]
```diff
@@ -43,34 +45,71 @@ class FitAndProperController @Inject()(
...
-              Redirect(routes.DetailedAnswersController.get(index, flow))
-          } recoverWith {
-            case _: IndexOutOfBoundsException => Future.successful(NotFound(notFoundView))
+            case ValidForm(_, data) => {
+              for {
+                cacheMap <- fetchAllAndUpdateStrict[ResponsiblePerson](index) { (_, rp) =>
+                  (appConfig.phase2ChangesToggle, data) match {
+                    case (true, true) => rp.approvalFlags(rp.approvalFlags.copy(hasAlreadyPassedFitAndProper = Some(data),
+                      hasAlreadyPaidApprovalCheck = Some(data)))
+                    case (true, false) => rp.approvalFlags(ApprovalFlags(hasAlreadyPassedFitAndProper = Some(data)))
```
Comment (slavvco) [inline app/controllers/responsiblepeople/FitAndProperController.scala]: why not copying here?

File: app/controllers/responsiblepeople/FitAndProperController.scala [hunk 2]
```diff
@@ -43,34 +45,71 @@ class FitAndProperController @Inject()(
...
+    (edit, fitAndProperAnswer, appConfig.phase2ChangesToggle) match {
+      case (_, _, false) => Redirect(routes.DetailedAnswersController.get(index, flow))        // Pre-phase 2             => Check your ans
+      case (true, false, true) => routeMsbOrTcsb(index, cacheMapOpt, flow)                     // Edit, no F&P, Phase 2   => Route based on business matching
+      case (true, true, _) => Redirect(routes.DetailedAnswersController.get(index, flow))      // Edit, F&P, Either phase => Check your ans
+      case (false, true, true) => Redirect(routes.DetailedAnswersController.get(index, flow))  // Create, F&P, Phase 2    => Check your ans
+      case (false, false, true) => routeMsbOrTcsb(index, cacheMapOpt, flow)                    // Create, no F&P, Phase 2 => Route based on business matching
+    }
+  }
+
+  private def routeMsbOrTcsb(index: Int, cacheMapOpt: Option[CacheMap], flow: Option[String]) :Result = {
```
Comment (slavvco) [inline app/controllers/responsiblepeople/FitAndProperController.scala]: routeMsbOrTcsp

File: app/controllers/responsiblepeople/FitAndProperController.scala [hunk 3]
```diff
@@ -43,34 +45,71 @@ class FitAndProperController @Inject()(
...
-        }
     }
+  }
 
-}
\ No newline at end of file
+  private def identifyRoutingTarget(index: Int, edit: Boolean, cacheMapOpt: Option[CacheMap],
+                                    flow: Option[String], fitAndProperAnswer: Boolean): Result = {
+    (edit, fitAndProperAnswer, appConfig.phase2ChangesToggle) match {
+      case (_, _, false) => Redirect(routes.DetailedAnswersController.get(index, flow))        // Pre-phase 2             => Check your ans
```
Comment (JGRobinson) [inline app/controllers/responsiblepeople/FitAndProperController.scala]: If its edit don't you always go back to check your answers?

File: app/controllers/responsiblepeople/FitAndProperController.scala [hunk 4]
```diff
@@ -43,34 +45,71 @@ class FitAndProperController @Inject()(
     implicit authContext => implicit request =>
 
       getData[ResponsiblePerson](index) map {
-        case Some(ResponsiblePerson(Some(personName),_,_,_,_,_,_,_,_,_,_,_,_,_,_,alreadyPassed,_,_,_,_,_,_)) if (alreadyPassed.hasAlreadyPassedFitAndProper.isDefined) =>
-          Ok(views.html.responsiblepeople.fit_and_proper(Form2[Boolean](alreadyPassed.hasAlreadyPassedFitAndProper.get), edit, index, flow, personName.titleName, appConfig.showFeesToggle, appConfig.phase2ChangesToggle))
+        case Some(ResponsiblePerson(Some(personName),_,_,_,_,_,_,_,_,_,_,_,_,_,_,alreadyPassed,_,_,_,_,_,_))
+          if (alreadyPassed.hasAlreadyPassedFitAndProper.isDefined) =>
```
Comment (slavvco) [inline app/controllers/responsiblepeople/FitAndProperController.scala]: parentheses not needed

File: app/controllers/responsiblepeople/FitAndProperController.scala [hunk 5]
```diff
@@ -43,34 +45,71 @@ class FitAndProperController @Inject()(
...
+      case (false, true, true) => Redirect(routes.DetailedAnswersController.get(index, flow))  // Create, F&P, Phase 2    => Check your ans
+      case (false, false, true) => routeMsbOrTcsb(index, cacheMapOpt, flow)                    // Create, no F&P, Phase 2 => Route based on business matching
+    }
+  }
+
+  private def routeMsbOrTcsb(index: Int, cacheMapOpt: Option[CacheMap], flow: Option[String]) :Result = {
+    cacheMapOpt match {
+      case Some(cacheMap) => {
+        (cacheMap.getEntry[BusinessMatching](BusinessMatching.key)) match {
+          case (Some(BusinessMatching(_, Some(BusinessActivities(acts, _, _, _)), _, _, _, _, _, _, _))) =>
```
Comment (slavvco) [inline app/controllers/responsiblepeople/FitAndProperController.scala]: parentheses not needed

File: app/controllers/responsiblepeople/FitAndProperController.scala [hunk 6]
```diff
@@ -43,34 +46,82 @@ class FitAndProperController @Inject()(
...
+                  cacheMap <- maybeCache
+                  bm <- cacheMap.getEntry[BusinessMatching](BusinessMatching.key)
+                } yield bm
+
+                val msbOrTcsp = ControllerHelper.isMSBSelected(Some(businessMatching)) ||
+                  ControllerHelper.isTCSPSelected(Some(businessMatching))
+
+                for {
+                  cacheMap <- fetchAllAndUpdateStrict[ResponsiblePerson](index) { (_, rp) =>
+                    if (appConfig.phase2ChangesToggle) {
```
Comment (JGRobinson) [inline app/controllers/responsiblepeople/FitAndProperController.scala]: Is this toggle the correct way round ?

Comment (JGRobinson) [reply app/controllers/responsiblepeople/FitAndProperController.scala]: sorry my bad - see it now

File: conf/application.conf [hunk 7]
```diff
@@ -115,7 +115,7 @@ microservice {
       show-fees = false
       enrolment-store = true
       fx-enabled = true
-      phase-2-changes = false
+      phase-2-changes = true
```
Comment (JGRobinson) [inline conf/application.conf]: this needs to be turned back off

Comment (JGRobinson) [reply conf/application.conf]: understand needs to be left on.

**Pull request merged**